processing/src/main/java/io/druid/segment/serde/DictionaryEncodedColumnPartSerde.java

```diff
@@ -19,66 +19,266 @@
+            NO_FLAGS,
+            dictionary,
+            singleValuedColumn.asWriteableSupplier(),
+            null,
+            bitmapSerdeFactory,
+            bitmaps,
+            spatialIndex,
+            byteOrder
+        );
+      } else if(multiValuedColumn != null) {
+        return new DictionaryEncodedColumnPartSerde(
+            VERSION.UNCOMPRESSED_MULTI_VALUE,
+            Feature.MULTI_VALUE.getMask(),
+            dictionary,
+            null,
+            multiValuedColumn,
+            bitmapSerdeFactory,
+            bitmaps,
+            spatialIndex,
+            byteOrder
+        );
+      } else {
+        throw new ISE("One of singleValuedColumn, compressedSingleValuedColumn, or multiValuedColumn must be set");
+      }
+    }
+  }
+
+  public static Builder builder()
+  {
+    return new Builder();
+  }
+
   private final BitmapSerdeFactory bitmapSerdeFactory;
+  private final ByteOrder byteOrder;
 
   private final GenericIndexed<String> dictionary;
-  private final VSizeIndexedInts singleValuedColumn;
+  private final WritableSupplier<IndexedInts> singleValuedColumn;
   private final VSizeIndexed multiValuedColumn;
   private final GenericIndexed<ImmutableBitmap> bitmaps;
   private final ImmutableRTree spatialIndex;
-
+  private final int flags;
+  private final VERSION version;
   private final long size;
 
+
+  @JsonCreator
   public DictionaryEncodedColumnPartSerde(
+      @JsonProperty("bitmapSerdeFactory") BitmapSerdeFactory bitmapSerdeFactory,
+      @JsonProperty("byteOrder") ByteOrder byteOrder
+  )
+  {
+    this.bitmapSerdeFactory = bitmapSerdeFactory == null
+                              ? new BitmapSerde.LegacyBitmapSerdeFactory()
+                              : bitmapSerdeFactory;
+    this.byteOrder = byteOrder;
+
+    this.dictionary = null;
+    this.singleValuedColumn = null;
+    this.multiValuedColumn = null;
+    this.bitmaps = null;
+    this.spatialIndex = null;
+    this.size = -1;
+    this.flags = 0;
+    this.version = VERSION.COMPRESSED;
+  }
+
+  private DictionaryEncodedColumnPartSerde(
+      VERSION version,
+      int flags,
       GenericIndexed<String> dictionary,
-      VSizeIndexedInts singleValCol,
-      VSizeIndexed multiValCol,
+      WritableSupplier<IndexedInts> singleValuedColumn,
+      VSizeIndexed multiValuedColumn,
       BitmapSerdeFactory bitmapSerdeFactory,
       GenericIndexed<ImmutableBitmap> bitmaps,
-      ImmutableRTree spatialIndex
+      ImmutableRTree spatialIndex,
+      ByteOrder byteOrder
   )
   {
-    this.isSingleValued = multiValCol == null;
+    Preconditions.checkArgument(singleValuedColumn != null ^ multiValuedColumn != null,
+                                "Exactly one of singleValCol[%s] or multiValCol[%s] must be set",
+                                singleValuedColumn, multiValuedColumn);
+
+    switch (version) {
+      case UNCOMPRESSED_MULTI_VALUE:
+        Preconditions.checkArgument(Feature.MULTI_VALUE.isSet(flags),
```

Comment:
Given that this constructor is private and the only entry point is the Builder, is this check important?  Or, can the code just enforce it?